### DISCUSSION TITLE
Task 14: iterative Weight Windows via Python API

### DISCUSSION THREAD

**Discussion by kingyue737** [fusion-energy/neutronics-workshop#274] [Q&A] [answered]
Firstly, thanks very much for this awesome repo. I couldn't find any example related to the new Weight Windows feature of OpenMC before reaching the repo here. This workshop provides a very complete tutorial.

The task 3 and 4 create iteratively a weight window to accelerate deep shielding simulations via CAPI. Can I achieve the same effect via PythonAPI to simplify my script? I notice that the task 2 didn't use `max_realizations` and `update_interval` of the new [ `openmc.WeightWindowGenerator`](https://docs.openmc.org/en/latest/pythonapi/generated/openmc.WeightWindowGenerator.html). I doesn't understand the description of these two attributes in the documentation. Are they used for iterating the weight window?

**Comment by shimwell** [comment, marked as answer]
Thanks for your kind comment and also for the PR fix for those typos earlier.

I did have a go at doing the same thing with the pure python api and you can do some weight windows. It has a disadvantage compared to the openmc.lib approach as openmc lib can build on the current simulation, avoid reloading nuclear data and ends up being a bit quicker.

The repo I made still exists so feel free to look at the source code over here https://github.com/fusion-energy/openmc_weight_window_generator


- max_realizations 
   - what the docs say max_realizations ([int](https://docs.python.org/3/library/functions.html#int)) – The upper limit for number of tally realizations when generating weight windows.
   - my interpetation. When running the openmc.WeightWindowGenerator you might want it to stop after a given number of batches even if it has not reached the converged std dev. So this can be used to set an upper limit to batches 

- update_interval 
   - what the docs say update_interval ([int](https://docs.python.org/3/library/functions.html#int)) – The number of tally realizations between updates.
   - my interpretation. when getting the weight window particles are run and then this is used to find the weight window. then another batch is run and the weight window can be used for this batch and that should normally help make the weight window better faster. But in some cases one might want to run a batch of particles and only update the weight window every 2nd or 3rd batch. In large weight window meshes the updating of the weight window can take time so it might be better to update it less frequently. Combinations of particles, batches and update_interval can be tweaked for different models and different optimal configurations may exist for different models or tallies.
